File: BuyAndSellStocks-II.py
"""
Time Complexity : O(n) number of elements in prices array
Space Complexity : O(1)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        smaller = prices[0]
        larger = prices[0]
        i = 0
        # We traverse through the prices and when ever we find a smaller element we
        # store the consecutive smaller and larger element in two variables
        # smaller and larger. We do it till for a particular smaller prices we find
        # the larger price. After that we just take the difference of larger and 
        # smaller element and add it to the maxProfit
        while i < len(prices) - 1:
            while i < len(prices) - 1 and prices[i] >= prices[i + 1]:
                i += 1
            smaller = prices[i]
            while i < len(prices) - 1 and prices[i] <= prices[i + 1]:
                i += 1
            larger = prices[i]
            maxProfit += larger - smaller
        return maxProfit

File: buyandsellstocks_ii.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices == null || prices.length == 0) return 0; //if the prices array is null or the length of the prices is 0 we return 0;
        int profit = 0; //we initialize an integer variable with name profit
        for(int i = 1; i < prices.length; i++) { //we start iterating from i = 1 until the length of the array of prices
            if(prices[i] > prices[i - 1]) { //if the price of stock at i is greater than the price of stock at i - 1
                profit += prices[i] - prices[i - 1]; // we add the difference to the profit variable
            }
        }
        return profit; // in the end, we return the profit.
    }
}
//tc and sc - O(n) and O(1)